packages/suite/src/actions/suite/metadataLabelingActions.ts

```diff
@@ -529,6 +568,297 @@ export const addMetadata =
+            if (!response.success) {
+                dispatch(
+                    metadataProviderActions.handleProviderError({
+                        error: response,
+                        action: ProviderErrorAction.LOAD,
+                        clientId: providerInstance.clientId,
+                    }),
+                );
+
+                return;
+            }
+
+            // todo: imho [] should be default return, it should not be also nullable
+            return response?.payload || [];
+        });
+
+        // no files, fresh account, no metadata encryption version migration needed
+        if (!files?.length) {
+            return [[], [], []];
+        }
+
+        // todo: this is not future proof in case there is another encryption version
+        const [currentEncryptionFiles, restFiles] = arrayPartition(files, file =>
+            file.endsWith(`_v2.mtdt`),
+        );
+
+        const [renamedOldEncryptionFiles, oldEncryptionFiles] = arrayPartition(restFiles, file =>
+            file.endsWith(`_v1.mtdt`),
+        );
+
+        return [currentEncryptionFiles, oldEncryptionFiles, renamedOldEncryptionFiles];
+    };
+
+const createMigrationPromise =
+    (
+        entity: LabelableEntity,
+        prevEncryptionVersion: MetadataEncryptionVersion,
+        fetchData: boolean,
+        device: TrezorDevice,
+    ) =>
+    async (dispatch: Dispatch, getState: GetState) => {
+        const encryptionVersion = selectEncryptionVersion(getState());
+
+        if (!device?.state || !device.metadata[encryptionVersion]) {
+            return { success: false, error: 'device unexpected state' };
+        }
+
+        let selectedProvider = selectSelectedProviderForLabels(getState());
+
+        const prevData =
+            fetchData &&
+            (await dispatch(
+                fetchMetadata({
+                    entity,
+                    encryptionVersion: prevEncryptionVersion,
+                    provider: selectedProvider!,
+                }),
+            ));
+
+        const nextKeys = entity[encryptionVersion];
+
+        if (!nextKeys) {
+            return { success: false, error: 'next keys are missing' };
+        }
+
+        const dummy = { dummy: getWeakRandomId(getRandomNumberInRange(1, 100)) };
+        const prevKeys = entity[prevEncryptionVersion];
+        if (!prevKeys) {
+            return { success: false, error: 'prev keys are missing' };
+        }
+
+        const defaultEntityData =
+            entity.type === 'account'
+                ? cloneObject(METADATA_LABELING.DEFAULT_ACCOUNT_METADATA)
+                : cloneObject(METADATA_LABELING.DEFAULT_WALLET_METADATA);
+
+        const nextData =
+            prevData && 'data' in prevData
+                ? { ...prevData.data, migratedFrom: prevKeys.fileName }
+                : { ...defaultEntityData, ...dummy };
+
+        selectedProvider = selectSelectedProviderForLabels(getState());
```

Comment:
We are getting fresh provider here. 

I think we shall split this function into multiple smaller parts with proper naming so it is clear how are the steps followed.